inmem.go

```diff
@@ -0,0 +1,244 @@
+package clover
+
+import (
+	"sort"
+	"sync"
+)
+
+type collection map[string]*Document
+
+type inMemEngine struct {
+	sync.RWMutex
+	collections map[string]collection
+}
+
+func newInMemoryStoreEngine() StorageEngine {
+	return &inMemEngine{
+		RWMutex:     sync.RWMutex{},
+		collections: make(map[string]collection),
+	}
+}
+
+func InMemoryDB() (*DB, error) {
+	return &DB{
+		dir:    "",
+		engine: newInMemoryStoreEngine(),
+	}, nil
+}
+
+// Close implements StorageEngine
+func (*inMemEngine) Close() error {
+	return nil
+}
+
+// CreateCollection implements StorageEngine
+func (e *inMemEngine) CreateCollection(name string) error {
+	e.Lock()
+	defer e.Unlock()
+	if ok, _ := e.HasCollection(name); ok {
+		return ErrCollectionExist
+	}
+	e.collections[name] = make(collection)
+	return nil
+}
+
+// Delete implements StorageEngine
+func (e *inMemEngine) Delete(q *Query) error {
+	e.Lock()
+	defer e.Unlock()
+
+	c, ok := e.collections[q.collection]
+	if !ok {
+		return ErrCollectionNotExist
+	}
+
+	for key, d := range c {
+		if q.satisfy(d) {
+			delete(c, key)
+		}
+	}
+
+	return nil
+}
+
+// DeleteById implements StorageEngine
+func (e *inMemEngine) DeleteById(collectionName string, id string) error {
+	e.Lock()
+	defer e.Unlock()
+
+	c, ok := e.collections[collectionName]
+	if !ok {
+		return ErrCollectionNotExist
+	}
+
+	delete(c, id)
+
+	return nil
+}
+
+func (e *inMemEngine) hasCollection(collectionName string) bool {
+	_, ok := e.collections[collectionName]
+	return ok
+}
+
+// DropCollection implements StorageEngine
+func (e *inMemEngine) DropCollection(name string) error {
+	e.Lock()
+	defer e.Unlock()
+
+	if ok, _ := e.HasCollection(name); ok {
+		delete(e.collections, name)
+	} else {
+		return ErrCollectionNotExist
+	}
+	return nil
+}
+
+// FindAll implements StorageEngine
+func (e *inMemEngine) FindAll(q *Query) ([]*Document, error) {
+	docs := []*Document{}
+	err := e.IterateDocs(q, func(doc *Document) error {
+		if q.satisfy(doc) {
+			docs = append(docs, doc)
+		}
+		return nil
+	})
+
+	return docs, err
+}
+
+// FindById implements StorageEngine
+func (e *inMemEngine) FindById(collectionName string, id string) (*Document, error) {
+	e.RLock()
+	defer e.RUnlock()
+
+	c, ok := e.collections[collectionName]
+	if !ok {
+		return nil, ErrCollectionNotExist
+	}
+
+	return c[id], nil
+}
+
+// HasCollection implements StorageEngine
+func (e *inMemEngine) HasCollection(name string) (bool, error) {
+	_, ok := e.collections[name]
+	return ok, nil
+}
+
+// Insert implements StorageEngine
+func (e *inMemEngine) Insert(collection string, docs ...*Document) error {
+	e.Lock()
+	defer e.Unlock()
+
+	c, ok := e.collections[collection]
+	if !ok {
+		return ErrCollectionNotExist
+	}
+
+	for _, d := range docs {
+		c[d.ObjectId()] = d
+	}
+
+	return nil
+}
+
+func (e *inMemEngine) iterateDocsSlice(q *Query, consumer docConsumer) error {
```

Comment:
This should be renamed to `iterateDocs`. In the badger engine, we have two private iterate functions, `iterateDocs()` and `iterateDocsSlice()` (this one is used when sorting is enabled). Here, we only have one, so it is better to simplify the name